### PR TITLE
FM2-157: Update PersonDAO and Service

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/FM2-157

Since i encountered new changes and it let my current PR  to have unneccesary merge conflicts i decided to create this PR to make it more clear 
 cc @ibacher  @robbyoconnor  @tendomart

### CODE REVIEW

**Comment by robbyoconnor** [comment]
@sherrif10 why did you tag me? :confused:

**Comment by robbyoconnor** [comment]
Also, [your build failed](https://travis-ci.com/github/openmrs/openmrs-module-fhir2/builds/164567819)

**Comment by sherrif10** [comment]
Yes please , am having wierd travis failure which i didnot expect

**Comment by dkayiwa** [comment]
@robbyoconnor he is missing you 😊

**Comment by sherrif10** [comment]
haaa ,Enway am having an error that  could arose from FhirPersonDaoImpl

**Comment by robbyoconnor** [comment]
> @robbyoconnor he is missing you 😊 

@dkayiwa not even sure I've been around much

**Comment by sherrif10** [comment]
This class is blocking me /FhirPersonDaoImpl. cc @ibacher , @dkayiwa  @robbyoconnor

**Comment by ibacher** [comment]
@sherrif10 So the error is caused because of a change in the class hierarchy. You've changed the `FhirPersonDaoImpl` class to extends `BaseFhirDao` instead of `BasePersonDao`. The problem is that the `getSqlAlias()` (declared on line 69) has the `@Override` annotation which raises a compiler error if it doesn't override a method defined in either the super class or an interface. Since the `getSqlAlias()` function is declared only in the `BasePersonDao` (correctly), you get the compiler error you're seeing.

So, to fix this, you need to understand that the `BasePersonDao` class is designed to implement functionality common to all of the classes that extend the OpenMRS `Person` type. That means that, in this case, we don't want to *replace* `BasePersonDao` with `BaseFhirDao`. Instead we want to *change* the `BasePersonDao` class to extend the `BaseFhirDao` class.

This is quite similar to all of the other changes in so far as what we're doing is inserting the `BaseFhirDao` into the super-class hierarchy before the `BaseDao` class.

Hope that clarifies things for you. Note that by changing the `BasePersonDao` you may need to make changes to other classes that extend the `BasePersonDao`, so you might actually solve several tickets with one PR.

**Comment by ibacher** [comment]
@sherrif10 So now the failures are:

```
[ERROR] /home/travis/build/openmrs/openmrs-module-fhir2/api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java:[92,30] cannot find symbol

  symbol:   method setFhirPersonDao(org.openmrs.module.fhir2.api.dao.FhirPersonDao)

  location: variable personService of type org.openmrs.module.fhir2.api.impl.FhirPersonServiceImpl

[ERROR] /home/travis/build/openmrs/openmrs-module-fhir2/api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java:[93,30] cannot find symbol

  symbol:   method setPersonTranslator(org.openmrs.module.fhir2.api.translators.PersonTranslator)

  location: variable personService of type org.openmrs.module.fhir2.api.impl.FhirPersonServiceImpl
```

This should be pretty straight-forward to fix. Inside the `FhirPersonServiceImplTest` you need to change the call to `setFhirPersonDao()` to use the new `setDao()` function and the `setPersonTranslator()` to use the new `setTranslator()` function.

**Comment by sherrif10** [comment]
Thanks @ibacher  it finally  travis passed,in same class i was thinking to make the field FhirPersonDao to dao, and fhirPersonTranslator to translator  but i think its very ok

File: omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java [hunk 1]
```diff
@@ -1,3 +1,12 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
```
Comment (corneliouzbett) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java:9]: There's a duplication of the license header. Also, this file shouldn't be in this PR.

Comment (sherrif10) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java]: Finished thanks

**Comment by sherrif10** [comment]
Ready for merge , i think at this level unless otherwise

**Comment by tendomart** [comment]
> https://issues.openmrs.org/browse/FM2-157
> 
> Since i encountered new changes and it let my current PR to have unneccesary merge conflicts i decided to create this PR to make it more clear
> cc @ibacher @robbyoconnor @tendomart

Thanks @sherrif10

**Comment by sherrif10** [comment]
Your welcome @tendomart

**Comment by corneliouzbett** [comment]
@sherrif10 one last thing remove this  `ObservationFhirResourceProviderTest` from your PR

**Comment by corneliouzbett** [comment]
> @sherrif10 one last thing remove this `ObservationFhirResourceProviderTest` from your PR

Resolved

**Comment by sherrif10** [comment]
That one is a minor one i tried to remove if but unfortunately it remained there, Could you merge it and we solve merge conflicts.Its ready for merge am sure, and its blocking me somewhere.

**Comment by sherrif10** [comment]
Any more idea please cc @ibacher  @corneliouzbett

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java [hunk 2]
```diff
@@ -17,38 +17,42 @@
...
 @Component
 @Transactional
+@Getter(AccessLevel.PROTECTED)
 @Setter(AccessLevel.PACKAGE)
-public class FhirPersonServiceImpl implements FhirPersonService {
+public class FhirPersonServiceImpl extends BaseFhirService<Person, org.openmrs.Person> implements FhirPersonService {
 	
 	@Autowired
-	private FhirPersonDao fhirPersonDao;
+	private FhirPersonDao Dao;
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java:39]: This should be `dao`.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java [hunk 3]
```diff
@@ -17,38 +17,42 @@
...
-	private PersonTranslator personTranslator;
+	private PersonTranslator translator;
 	
 	@Override
 	@Transactional(readOnly = true)
-	public Person getPersonByUuid(String uuid) {
-		return personTranslator.toFhirResource(fhirPersonDao.getPersonByUuid(uuid));
+	public Person get(String uuid) {
+		return translator.toFhirResource(Dao.get(uuid));
 	}
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java:48]: You should be able to remove this function altogether

File: api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java [hunk 4]
```diff
@@ -79,7 +79,7 @@
 	private org.openmrs.Person person;
 	
 	@Mock
-	private FhirPersonDao dao;
+	private FhirPersonDao fhirPersonDao;
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java:82]: Why did you rename this?

File: omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java [hunk 5]
```diff
@@ -3,7 +3,6 @@
  * v. 2.0. If a copy of the MPL was not distributed with this file, You can
  * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
  * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
- *
```
Comment (ibacher) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java:6]: Please remove this from this PR.

File: api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java (hunk 4 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java:82]: This is a vairable except if am to change it where it was used. but by default it was like that

File: omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java (hunk 5 above)
Comment (sherrif10) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java]: Finished

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java (hunk 3 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java]: Finished

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java (hunk 2 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImpl.java]: Finished

File: omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java (hunk 5 above)
Comment (sherrif10) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/ObservationFhirResourceProviderTest.java]: Anything missing @ibacher @corneliouzbett

File: api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java (hunk 4 above)
Comment (ibacher) [reply api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirPersonServiceImplTest.java]: I'd just leave this as `dao`.
